### PR TITLE
Refactor player update handling + scene calling

### PR DESCRIPTION
**Pull request by fmatthew5876**
~~Depends on #1697~~ 

Fix: #1691 
Fix: #1457 

Fixes the gameover part of #1642. The message vs gameover stuff is not fixed yet, but that looks like a message bug to be fixed in another PR.

I also haven't studied too deeply the immediate "to title" behavior.

2 more game temp variables killed!

### CODE REVIEW

File: src/game_interpreter.cpp [hunk 1]
```diff
@@ -1540,8 +1542,9 @@ bool Game_Interpreter::CommandGameOver(RPG::EventCommand const& /* com */) { //
 	if (Game_Message::visible) {
 		return false;
 	}
-	Game_Temp::gameover = true;
-	SetContinuation(&Game_Interpreter::DefaultContinuation);
+
+	Scene::instance->SetRequestedScene(Scene::Gameover);
+	++index;
```
Comment (Ghabry) [inline src/game_interpreter.cpp]: > Refactor Gameover handling 

What's this new ++index doing here?
Is that for aborting Game Over calls when another scene call overwrites in the same frame?

File: src/game_player.cpp [hunk 2]
```diff
@@ -218,8 +258,21 @@ void Game_Player::UpdateSelfMovement() {
...
+				}
+			}
+		}
+	}
+
+	// ESC-Menu calling
+	if (Game_System::GetAllowMenu()
+			&& !Game_Message::visible
+			&& !Game_Map::GetInterpreter().IsRunning())
+	{
```
Comment (Ghabry) [inline src/game_player.cpp]: > Refactor player update routine order 

I'm not sure about the change for ESC-Menu calling.

I thought the menu can't be accessed when the route is overwritten or a message is pending to be shown but both these conditions were removed.

File: src/game_interpreter.cpp (hunk 1 above)
Comment (fmatthew5876) [reply src/game_interpreter.cpp]: That's right, otherwise we'd be stuck calling it over and over.

File: src/game_player.cpp (hunk 2 above)
Comment (fmatthew5876) [reply src/game_player.cpp]: `UpdateSelfMovement()` is not called by `Game_Character` if `IsMoveRouteOverwritten() == true`. 

For the message pending part, I'll need to check how that overlaps with visible.

Comment (fmatthew5876) [reply src/game_player.cpp]: It's confusing as we don't have a standard way to ask "Is there a message active?"

`message_waiting` goes from when the `ShowMessage` command was invoked up until the message box starts to close. `visible` goes from the first frame of the actual message until the message close animation is done. So in most cases you probably want both.

I changed this to check for both. Also I added back a check for `!IsMoveRouteOverwritten()` I guess in theory if you press a direction and ESC at the same time, the player directional move could cause an event to run, which then sets a move route, which then prevents the player from opening the menu in the same frame...

**Comment by Ghabry** [comment]
Has deps so not mergeable but except for the menu calling and message waiting which I find suspicious and will do more testing with I give my :+1: here.

> Refactor player update routine order based on Cherry

Can you find the post where Cherry explained this?

**Comment by fmatthew5876** [comment]
More info here https://github.com/EasyRPG/Player/issues/1691

My original conversation with Cherry is here:
https://github.com/EasyRPG/Player/issues/1626#issuecomment-469976087

**Comment by fdelapena** [comment]
Diff between PR1697 and PR1714: https://github.com/fmatthew5876/Player/compare/event_trigger...fmatthew5876:player_scene

**Comment by fmatthew5876** [comment]
jenkins: test this please

**Comment by Ghabry** [comment]
hm the moving is really blocked until the box is closed, not until it starts closing. Nobody noticed this for 10 years.

File: src/game_player.cpp [hunk 3]
```diff
@@ -185,9 +185,50 @@ void Game_Player::UpdateScroll(int old_x, int old_y) {
...
 void Game_Player::UpdateSelfMovement() {
-	if (!Game_Map::GetInterpreter().IsRunning() && !Game_Map::IsAnyEventStarting()) {
-		if (IsMovable()) {
+	if (!IsBoardingOrUnboarding()
+			&& !Game_Map::GetInterpreter().IsRunning()
+			&& !Game_Message::message_waiting
+			&& !Game_Message::visible
+			&& !IsMoveRouteOverwritten()
+			&& !IsPaused() // RPG_RT compatible logic, but impossible to set pause on player
+			&& !Game_Map::IsAnyEventStarting())
```
Comment (Ghabry) [inline src/game_player.cpp]: imo this could be in "IsMovable". When you look into IsMovable almost all of this is already in there (Graphics::TransitionPending is nonsense and can be left out).
Also IsMovable is currently dead code, nothing calls it.
IsMovable even contains a case you forgot:

bla

(fancy APNG, should be animated)

**Comment by Ghabry** [comment]
also when the airship is asc/desc'ing the menu is delayed until the animation finishes, in Player is immediate, not a regression, just for the notes.

File: src/game_player.cpp (hunk 3 above)
Comment (fmatthew5876) [reply src/game_player.cpp]: Nice catch! Thanks

**Comment by fmatthew5876** [comment]
I fixed the vehicle issues, including the menu thing.

Those are good catches, thanks for that.

I opted to just remove `IsMovable()` since it isn't really applicable anywhere else. Keeps the logic localized in one place and easier to follow imo.

**Comment by fmatthew5876** [comment]
Test 4 in #1691 proves we need to check `message_waiting`

If I remove the `message_waiting` check, player will print "hello" and then call the menu after the message box dissapears.

**Comment by fmatthew5876** [comment]
It looks like Ghabry's original concerns about `IsMoveRouteOvewritten()` were valid. You can press ESC during move route and menu will be summoned when the move route finishes.

Also more edges cases related to this are fixed, and described in test cases.

**Comment by fmatthew5876** [comment]
jenkins: test this please

**Comment by Ghabry** [comment]
Oh, so the vehicle menu calling was just an edge case, good finds :+1:
